first cut at a CSV reader for Spotter buoys

ryancoe opened the pull request:
Closes #66

ryancoe commented on tests/io/test_spotter.py:
```diff
@@ -32,3 +34,30 @@ def test_stats(self, wavespectra_name, spotter_name, kwargs):
+#     )
+#     def test_stats(self, wavespectra_name, spotter_name, kwargs):
+#         """Assert that stat calculated from wavespectra matches the one read from spotter."""
+#         wavespectra = getattr(self.dset.spec, wavespectra_name)(**kwargs).values
```
I've never used the class-based pytest structure you're using here and I'm also having a hard time understanding these lines (I think that you are comparing bulk parameters reported in the Spotter file to those you calculate from the spectral density, but I don't quite get how this works, particularly where in the `read_spotter` `JSON` workflow the `.dset.spec` field is created): https://github.com/wavespectra/wavespectra/blob/1a5a02fbbea3ab332e8b9b14d44e90beac631f54/tests/io/test_spotter.py#L32-L33

rafa-guedes replied:
These tests use this `@pytest.mark.parametrize` to allow running the function that comes right after (`test_stats`) with a range of input argument combinations (which are defined in that list of tuples in the arg of parametrize  decorator).

`dset.spec` is not actually created by the json reader per se - but it is the `SpecArray` accessor that wavespectra itself defines when it is imported and it finds a dataset with appropriate conventions. You don't have to worry about it, the important thing to know is that you have to return the xarray dataset from your reader with the appropriate conventions (specified [here](https://wavespectra.readthedocs.io/en/latest/conventions.html)).

You can define your tests in a different way that makes sense to you, as long as the tests are robust and accesses as much as the functionality in your code as possible in order to test the code well. But you got the existing one basically right, it simply compares stats calculated from wavespectra against those defined in the json file.

ryancoe replied:
Oh cool! While I don't really understand how his happens (something [here](https://github.com/wavespectra/wavespectra/blob/master/wavespectra/specdataset.py#L37) I think), I think I understand how to interact with it now.

ryancoe commented on wavespectra/input/spotter.py:
```diff
@@ -6,16 +6,206 @@
+    
+    ds = ds.sortby('time')
+    
+    return ds
```
It looks like maybe I should separate data variables into `.dset` (which includes only the variables listed here: https://wavespectra.readthedocs.io/en/latest/conventions.html#data-variables) and `.spot` which contains anything else recorded in the `CSV` (e.g., `Battery Voltage`, `Significant Wave Height`, `Peak Period`, etc.)

rafa-guedes replied:
No, please see my comment above. You don't really touch the `.spec` accessor, that is created by the SpecArray / SpecDataset main classes themselves. All you have to do is return an xarray dataset with the correct conventions regarding coordinate naming and units.

ryancoe commented on wavespectra/input/spotter.py:
```diff
@@ -6,16 +6,206 @@
+    ds.attrs['references'] = 'https://content.sofarocean.com/hubfs/Technical_Reference_Manual.pdf'
+
+
+    ds = ds.rename({'Epoch Time':'time',
```
It looks like I should maybe set variable names and `attrs` using [`attributes.yml`](https://github.com/wavespectra/wavespectra/blob/master/wavespectra/core/attributes.yml) -- I see this is being done in the `JSON` workflow, but don't immediately understand how to do it in the workflow I'm using the for `CSV` files

rafa-guedes replied:
There is a function to help with attribute definition, it uses that attributes.yml as you correctly noted. You import it from here (it is already imported in spotter.py actually):

```
from wavespectra.core.attributes import set_spec_attributes
```

and then you just run something like:

```
set_spec_attributes(dset)
```

before returning dset in your reader function.

rafa-guedes replied:
However it looks like you also return several other non-spectral variables in the output - those won't have attributes defined in attributes.yml so you will have to manually define that for those as you are doing.

ryancoe replied:
Since I already do this manually and, as you note, include many other variables, I'm going to forgo the `set_spec_attributes(dset)` method unless that's an issue for you.

ryancoe commented on wavespectra/input/spotter.py:
```diff
@@ -6,16 +6,206 @@
+    return ds
+
+
 def read_spotter(filename):
```
Would probably be good (and not too hard) to more elegantly integrate the added `CSV` reading capability with the existing `JSON` workflow (e.g., via a simple `if-elif` statement as discussed here: https://github.com/wavespectra/wavespectra/issues/66#issuecomment-1292680831.

ryancoe commented on wavespectra/input/spotter.py:
```diff
@@ -6,16 +6,206 @@
 from wavespectra.core.attributes import attrs, set_spec_attributes
 
 
+def read_spotter_csv(filename):
```
would be nice to handle (and concatenate) multiple `CSV` files as is currently done in the `JSON` Spotter workflow

rafa-guedes commented:
Hi @ryancoe 

thanks for submitting this pull request and apologies for taking a bit long to get back to you, I have been a bit busy these last couple of weeks! This looks great. I will write a few comments in separate threads so we can discuss!

rafa-guedes commented:
## Reader function signature

Defining the new function called `read_spotter_csv` inside the `spotter.py` module could be done but it means the reader won't be imported at the main module level so we could for instance do: 

```from wavespectra import read_spotter_csv```

To do that we need to to name the function as something like `read_{module_name}` as explained [here](https://wavespectra.readthedocs.io/en/latest/io-input.html). We have some functions defined without this convention, for instance `read_hotswan` which is defined inside the `wavespectra.input.swan` module - but those were not intended to be imported at module level in any case as they are kind of specific.

In fact it could be nice having only one reader for spotter, and having an argument to specify whether we want to read data in `json` or in `csv` format, as you suggested in one of your comments above. We can maybe leave the two functions and rename them as `_read_spotter_csv`, `_read_spotter_json` and then add just a wrapper function `read_spotter` which handles the logic and is imported at the module level.

rafa-guedes commented:
## Code formatting

We require docstrings in a certain convention (we use something similar to the Google convention) in order to build the readthedocs documentation automatically - if you could add that please it would be great. Just use as a template some from an existing reader such as `read_spotter`.

Also, the code is formatted using a tool called `black`. You can pip install it and just run `black path/to/spotter.py`. It just formats your script using black's convention which is a subset from pep8. The good thing is it makes all the code in the library look consistent and you don't have to worry about code formatting when you are writing it!

rafa-guedes commented:
## Time coordinate

I had a look at some file read with the new reader and noticed there is `time` and `Epoch Time ` coordinates - and their values are not the same (in the test file, `time` starts at `2021-09-03T16:12:01.000000000` and `Epoch Time ` starts at `2021-09-28T13:12:01.000000000` Also, some variables use time and some Epoch Time as a coordinate. Do we need the two coordinates? Why do they have different values?

ryancoe commented:
@rafa-guedes - I think I've addressed all of the issues we've discussed. The resulting `spotter.py` module is definitely not as clean as it could be. This is partially driven by the `JSON` and `CSV` files providing different information and partially by me wanting to adapt my existing `CSV` code and not start from scratch. Anyhow, not sure if this is OK with you, but hopefully it provides something worthwhile.

rafa-guedes commented:
@ryancoe thanks a lot for your contribution, it looks great! This will be available in the next release `v3.13.0`. Apologies for taking a bit long, I wanted to include a few other changes in this next release as well.

rafa-guedes commented:
Implemented in https://github.com/wavespectra/wavespectra/tree/v3.13.0
